### ISSUE TITLE
Cannot process thousands of items in Feasible Time

### ISSUE THREAD

**Issue by bksubhuti** [dart-lang/language#2588]
I have a dictionary with 200,000 records.
Normally I process these records in js and I can read and write an 18MB file within a few seconds.  (Note:.. older file about 168,000).

However, when I tried doing this in Dart for our re-written product, I ran into problems with .fromJson parsing.  There seemed to be a limit of less than 90,000.

Further investigation had me try processing a csv manually.
Further investigation had me write a csv manually in case my data was bad.
Further investigation had me write a dart cli ...

The code is shown below and takes roughly 1 second per 1000 merely to do string processing in memory.
The flutter code of the same thing takes about 2.5 seconds per 1000.
So in flutter (for my real life dictionary import feature), we are talking about 600 seconds or 10 minutes.  I have not ran any of these scripts to completion.  They appear as "hung" even though they are just taking a long time.

Meanwhile.. I can read an sql file and pass db.Update(sql); in a few seconds.
For now, I will need to store my data online for downloading dictionary updates as Raw SQL transaction inserts.

I never thought I would report speed as a bug.  But when we are talking a few seconds in js versus 10 minutes, this appears as "hung" and counts as a bug.  I love dart flutter, but had I known about this limitation, I might have opted for something else.  on the other hand the framework and using db's does not require us to feel this pain until now.

Hardware: Lenovo IdeaPad 3 14ALC6
Ubuntu 22.04.1
[✓] Flutter (Channel stable, 3.3.5, on Ubuntu 22.04.1 LTS 5.15.0-52-generic, locale en_US.UTF-8)

```
import 'package:cli/cli.dart' as cli;
import 'dart:core';
```

```
void main(List<String> arguments) {
  //print('Hello world: ${cli.calculate()}!');

  String myString = "";
  // make 200k records
  for (int x = 0; x < 200000; x++) {
    myString += "$x, another variable$x, 8\n";
    if (x % 1000 == 1) {
      print("$x");
    }
  }
}

```

**Comment by blopker** [comment]
In Dart, string concatenation is a bit slow. I'd suggest using StringBuffer to build larger strings:

```
import 'dart:core';

void main(List<String> arguments) {
  var myString = StringBuffer();
  // make 200k records
  for (int x = 0; x < 200000; x++) {
    myString.writeln("$x, another variable$x, 8");
    if (x % 1000 == 1) {
      print("$x");
    }
  }
  print(myString.length);
}
```
This takes less than a second on my machine.

**Comment by bksubhuti** [comment]
Yes.. correct.  My friend also had another fix with some similar but different code.
These numbers are more reasonable, but the question is .. is my original code out of the question for what can be practiced.
Also.. I came about this problem while trying to parse a JSON file of 200k records.  This is another problem.
I wonder if the "+=" operator can be overloaded withe a write instead of writeln.

I think something should change.. 10 minutes versus a second .. yes.. that is a bug to me.
At a minimum.. lint should pick it up and suggest.

**Comment by blopker** [comment]
I'm just a Dart user, not on the team or anything, but it's pretty common to use a StringBuffer in other languages when building large strings. I guess I personally don't have an issue using it in these cases, which I've only ran into once. It is interesting why JS is so much faster using the + operator though.

**Comment by bksubhuti** [comment]
a string and a buffer should be the same, but i guess it is not.  My friend said that 200k objects are being made with this method.  It also gets cumulatively slower and slower as more records (concatenations) are made.  so each copy has more "string"

**Comment by Levi-Lesches** [comment]
> My friend said that 200k objects are being made with this method.

Yep, your friend's right: 
- `x += y` is equivalent to `x = x + y`. But, since Strings are not actually mutable, Dart has to allocate a _third_ string equal to `x + y` and replace `x` with that new string. 
- StringBuffer is more efficient. The docs don't have that many details, but they do say it avoids allocating a new string until you explicitly ask it to. That means `buffer.write(y)` does _not_ allocate a third string, but rather adds the characters from `y` to the `buffer`. This is more efficient in terms of space and time

> These numbers are more reasonable, but the question is .. is my original code out of the question for what can be practiced?

There is a lint that warns you not to use `+=` for strings: [`use_string_buffers`](https://dart-lang.github.io/linter/lints/use_string_buffers.html). It's not enabled by default though. I just opened an issue to ask to move it to the "on by default" lints: https://github.com/dart-lang/lints/issues/90. Funnily enough, this lint has code similar to yours as an example of what _not_ to do: 

```dart
String foo() {
  final buffer = '';
  for (int i = 0; i < 10; i++) {
    buffer += 'a'; // LINT
  }
  return buffer;
}
```
vs
```dart
String foo() {
  final buffer = StringBuffer();
  for (int i = 0; i < 10; i++) {
    buffer.write('a');
  }
  return buffer.toString();
}
```

> I wonder if the "+=" operator can be overloaded with a write instead of `writeln`?

Currently, `+=`, `++`, and all the self-mutating operators cannot be overridden. See #2580.

Hopefully that helps
